server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.bulk;
+
+import org.elasticsearch.telemetry.metric.LongCounter;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+
+import java.util.Map;
+
+/**
+ * A record APM metrics that concern failure stores.
+ */
+public record FailureStoreMetrics(LongCounter failureCounter, LongCounter redirectCounter) {
+
+    public static final FailureStoreMetrics NOOP = new FailureStoreMetrics(MeterRegistry.NOOP);
+
+    public static final String METRIC_FAILURE_TOTAL = "es.ingest.failures.total";
+    public static final String METRIC_REDIRECT_TOTAL = "es.ingest.failure_redirects.total";
+
+    public FailureStoreMetrics(MeterRegistry meterRegistry) {
+        this(
+            meterRegistry.registerLongCounter(METRIC_FAILURE_TOTAL, "ingest failure counter", "unit"),
+            meterRegistry.registerLongCounter(METRIC_REDIRECT_TOTAL, "ingest failure redirect counter", "unit")
+        );
+    }
+
+    public static void incrementForIndex(LongCounter counter, String index) {
```

Comment:
I wonder if perhaps we should capture if the failure is an ingest node failure or a shard level failure. Can we think of any other attributes that might be useful to capture?

Reply:
I also think this is a good metric to have.

If we track metrics per index, should we also track per pipeline?

Also, should we track per type of error? Initially, I would think that this would vary too much, but if we split the metrics by index name, then I guess the error type is not that crazy either.

Reply:
I'm inclined to agree that it makes sense to add all these attributes. My only question/concern is that I don't know how these metrics are stored exactly, and thus what the impact would be of adding attributes (mainly in terms of memory consumption I guess). Is either of you familiar with the inner workings of APM? If not, I can do some investigation.

Reply:
Sorry, I would not expect that counters would be that heavy, but we should understand how these metrics could grow, ballpark figures I mean.

Reply:
> Full disclosure, part of the reason I'm having second thoughts is that it's slightly inconvenient to get the error type and pipeline, as we have to perform some checks to able to obtain them

I'm alright with leaving ingest node specific stats off for now. If we find that we need that data we can try to add them to the attributes in a follow up.

> I'm not sure all these attributes are actually valuable. Wouldn't these attributes overlap with/overshadow a part of the purpose of the failure store? In other words, what do both of you see as the value of adding these attributes?

I think having a way to track these failure stats when managing multiple clusters would be beneficial. Failure stores could hold sensitive information in them, so operators and maintainers may not always have access or permission to view the failure store. Collecting metadata from multiple clusters about failure rates while avoiding identifying information so that it can be analyzed in a central place is likely worth the double counting.

> I wonder if perhaps we should capture if the failure is an ingest node failure or a shard level failure. Can we think of any other attributes that might be useful to capture?

I think capturing the following attributes would be beneficial:

* The data stream we are writing the data to
  * Lets us see what kinds of integrations are experiencing problems
* The kind of failure we're experiencing (ingest node|shard level)
* The exception type that we've encountered
  * This helps us identify the failure without exposing any sensitive information about the failed documents
* Whether or not the data stream we are writing the data to has failure stores enabled or not
  * Failure stores will likely never be on for every data stream by default and users will have the option of disabling them on integrations. This helps us identify where any gaps in our coverage may be, and which integrations may be dropping failed data on the floor.

In the future I could see the following be helpful to trace:
* Ingest pipeline name and the processor type+tag to help identify specific problems
* If our errors ever end up with a unique id for their cause we could capture that (like 
* Some day we could try and capture which field is causing the problem with mappings?

Reply:
I think what you're saying definitely makes sense to me. I do have some follow-up thoughts/questions:
- Do we still see value in tracking the number of failures that got redirected? 
- Does it make sense to _only_ track the number of failures for data streams that do not have failure stores enabled? Neither in the scenario you sketched nor in any I can think of right away do I see great value in tracking those numbers for data streams that do not have failure stores enabled. 
- As @gmarouli pointed out, it probably doesn't make sense to track these numbers for non-data stream failures at all, correct?

Reply:
> Do we still see value in tracking the number of failures that got redirected?

I think it's still valuable to collect these for the sake of transmitting them via APM. Otherwise we'll need to source it from something like the stats, and those stats may have different retention times than those we collect from APM.

> Does it make sense to only track the number of failures for data streams that do not have failure stores enabled? Neither in the scenario you sketched nor in any I can think of right away do I see great value in tracking those numbers for data streams that do not have failure stores enabled.

Tracking failed documents for data streams without the failure store enabled would allow us to identify how much data we're rejecting outright that could have been saved into a failure store. @felixbarny do you have strong feelings about this point?

> As @gmarouli pointed out, it probably doesn't make sense to track these numbers for non-data stream failures at all, correct?

Let's keep them to data stream failures only. Let's not track requests to standalone indices. Agreed there.